### PR TITLE
feat(ecs-patterns) expose multiple target groups, listeners, and load balancers 

### LINKED ISSUE #14735: (ecs-patterns): Expose arrays of load balancers, listeners, and targetGroups as properties in ApplicationMultipleTargetGroupsFargateService and other patterns constructs
I want to utilize the construct *ApplicationMultipleTargetGroupsFargateService* for creating a Fargate service fronted by an Application Load Balancer. I need the Application Load Balancer to have 2 listeners and each listener to forward traffic to a distinct ApplicationTargetGroup. 

The problem I run is after I instantiate the construct, and want to configure health checks for each individual target group. The *ApplicationMultipleTargetGroupsFargateService* provides a *targetGroup* property, however, according to the CDK documentation, this only contains a reference to the default Target Group. Ideally, this would contain a list of references to the different ApplicationTargetGroups configured within the construct. Is there any method for getting a reference to these additional target groups in the CDK from the construct for configuring health checks? 

```
const service = new ApplicationMultipleTargetGroupsFargateService(this, 'ALBFargateService', {
    cluster: this.ecsALBCluster,
    taskDefinition: taskDefinition,
    desiredCount,
    assignPublicIp: true, // needed for outbound network connectivity in a public subnet
    serviceName: 'ALBFargateService',
    loadBalancers: [
        {
            name: 'ApplicationLoadBalancer',
            listeners: [
                {
                    name: 'TrafficListener',
                    port: 443,
                    protocol: ApplicationProtocol.HTTP
                },
                {
                    name: 'HealthCheckListener',
                    port: 8050,
                    protocol: ApplicationProtocol.HTTP
                }
            ]
        }
    ],
    targetGroups: [
        {
            listener: 'TrafficListener',
            containerPort: 443,
            protocol: Protocol.TCP
        },
        {
            listener: 'HealthCheckListener',
            containerPort: 8050,
            protocol: Protocol.TCP
        }
    ]
});

service.targetGroup.configureHealthCheck({
    port: '8050',
    protocol: elbv2.Protocol.HTTP,
    healthyThresholdCount: 2,
    unhealthyThresholdCount: 2,
    timeout: Duration.seconds(10),
    interval: Duration.seconds(30),
    healthyHttpCodes: '200'
});

// TODO: Configure Health Checks for second target group on port 8050
```

As shown above, I am configuring the service with 2 listeners and 2 target groups associated with a respective listener. Only the first target group's reference is accessible via the targetGroup property.

### PR DESCRIPTION
**Pull request by michaeldrey**
This commit will expose targetGroups, listeners, and loadBalancers properties allowing developers access to those resources.

The downside to this is that now we have targetGroups, targetGroup, listeners, listener, and loadBalancers, loadBalancer exposes as a property. In my opinion this can cause confusion for developers.
 
** Potential breaking change **
When trying to protect targetGroup, listener, and loadBalancer I discovered that this would cause breaking changes on down developers who use them to reference `node.Id` and other attributes. 

Only thing I could do was mark them as `deprecated` as you get the same functionality on the new properties. Open to feedback on this approach and will update the README based on the direction we go.

I need to update the network Integ tests some more, but looking for feedback on the above

closes #14735
----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by TheRealAmazonKendra** [comment]
So, I have a couple things that I need clarification on. I keep coming back to this PR and feeling unsure of how to respond because I don't get why we didn't just make these arrays in the first place. Seems like a weird design decision there and that makes me think there had to be some reason for it? Mostly I'm concerned that I'm missing context here. I've read through the original PR for this feature and I'm kind of stumped.

You noted that this is a potentially breaking change but that by deprecating them it is not breaking. Am I understanding that correctly? If so, this is the route I think is correct.

**Comment by TheRealAmazonKendra** [comment]
In terms of keeping it from being a breaking change, I do like your approach. If you think it would be better to take the other approach, however, it could be done under a feature flag. I'm fine with whatever you think is the better path. Go ahead and assign this to me when it's ready for review (per your notes in the PR about needing to do some additional updates)

**Comment by TheRealAmazonKendra** [comment]
A big thanks to @mrgrain who did dug into this with me. It turns out that we released this before multiple target groups was supported. Given that, I like this current path and will re-review today.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts [hunk 1]
```diff
@@ -371,24 +371,36 @@ export abstract class ApplicationMultipleTargetGroupsServiceBase extends Constru
...
    */
   public readonly loadBalancer: ApplicationLoadBalancer;
 
   /**
-   * The default listener for the service (first added listener).
-   */
+    * The default listener for the service (first added listener).
+    *
+    * @deprecated - Use `listeners` instead.
+    */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:383]:
```suggestion
   * @deprecated - Use `listeners` instead.
   */
```

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:383]: Spacing got a little off in a number of places.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts [hunk 2]
```diff
@@ -371,24 +371,36 @@ export abstract class ApplicationMultipleTargetGroupsServiceBase extends Constru
...
   public readonly listener: ApplicationListener;
 
   /**
-   * The cluster that hosts the service.
-   */
+    * The cluster that hosts the service.
+    */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:388]:
```suggestion
   * The cluster that hosts the service.
   */
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts [hunk 3]
```diff
@@ -371,24 +371,36 @@ export abstract class ApplicationMultipleTargetGroupsServiceBase extends Constru
 
   /**
    * The default Application Load Balancer for the service (first added load balancer).
+   *
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:374]:
```suggestion
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts [hunk 4]
```diff
@@ -371,24 +371,36 @@ export abstract class ApplicationMultipleTargetGroupsServiceBase extends Constru
...
-  protected targetGroups = new Array<ApplicationTargetGroup>();
-
-  private loadBalancers = new Array<ApplicationLoadBalancer>();
+  /**
+    * The listeners of the service.
+    */
+  public readonly listeners = new Array<ApplicationListener>();
+  /**
+       * The target groups of the service.
+       */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:398]:
```suggestion
  * The target groups of the service.
  */
```

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:398]: I might have pushed this one too far back. I can't totally tell.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts [hunk 5]
```diff
@@ -371,24 +371,36 @@ export abstract class ApplicationMultipleTargetGroupsServiceBase extends Constru
...
+  public readonly targetGroups = new Array<ApplicationTargetGroup>();
+  /**
+       * The load balancers of the service.
+       */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts:402]:
```suggestion
  * The load balancers of the service.
  */
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts [hunk 6]
```diff
@@ -289,24 +289,36 @@ export abstract class NetworkMultipleTargetGroupsServiceBase extends Construct {
 
   /**
    * The Network Load Balancer for the service.
+   *
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:292]:
```suggestion
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts [hunk 7]
```diff
@@ -289,24 +289,36 @@ export abstract class NetworkMultipleTargetGroupsServiceBase extends Construct {
...
    */
   public readonly loadBalancer: NetworkLoadBalancer;
 
   /**
-   * The listener for the service.
-   */
+    * The listener for the service.
+    *
+    * @deprecated - Use `listeners` instead.
+    */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:301]:
```suggestion
   * The listener for the service.
   * @deprecated - Use `listeners` instead.
   */
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts [hunk 8]
```diff
@@ -289,24 +289,36 @@ export abstract class NetworkMultipleTargetGroupsServiceBase extends Construct {
...
   public readonly listener: NetworkListener;
 
   /**
-   * The cluster that hosts the service.
-   */
+    * The cluster that hosts the service.
+    */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:306]:
```suggestion
   * The cluster that hosts the service.
   */
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts [hunk 9]
```diff
@@ -289,24 +289,36 @@ export abstract class NetworkMultipleTargetGroupsServiceBase extends Construct {
...
+    * The listeners of the service.
+    */
+  public readonly listeners = new Array<NetworkListener>();
+  /**
+    * The target groups of the service.
+    */
+  public readonly targetGroups = new Array<NetworkTargetGroup>();
+  /**
+    * The load balancers of the service.
+    */
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:320]:
```suggestion
   * The listeners of the service.
   */
  public readonly listeners = new Array<NetworkListener>();
  /**
   * The target groups of the service.
   */
  public readonly targetGroups = new Array<NetworkTargetGroup>();
  /**
   * The load balancers of the service.
   */
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/ecs/application-multiple-target-groups-ecs-service.ts [hunk 10]
```diff
@@ -92,6 +92,8 @@ export class ApplicationMultipleTargetGroupsEc2Service extends ApplicationMultip
   public readonly taskDefinition: Ec2TaskDefinition;
   /**
    * The default target group for the service.
+   *
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/ecs/application-multiple-target-groups-ecs-service.ts:95]:
```suggestion
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/ecs/network-multiple-target-groups-ecs-service.ts [hunk 11]
```diff
@@ -91,6 +91,8 @@ export class NetworkMultipleTargetGroupsEc2Service extends NetworkMultipleTarget
   public readonly taskDefinition: Ec2TaskDefinition;
   /**
    * The default target group for the service.
+   *
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/ecs/network-multiple-target-groups-ecs-service.ts:94]:
```suggestion
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/fargate/application-multiple-target-groups-fargate-service.ts [hunk 12]
```diff
@@ -106,6 +106,8 @@ export class ApplicationMultipleTargetGroupsFargateService extends ApplicationMu
 
   /**
    * The default target group for the service.
+   *
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/fargate/application-multiple-target-groups-fargate-service.ts:109]:
```suggestion
```

File: packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-multiple-target-groups-fargate-service.ts [hunk 13]
```diff
@@ -106,6 +106,8 @@ export class NetworkMultipleTargetGroupsFargateService extends NetworkMultipleTa
 
   /**
    * The default target group for the service.
+   *
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ecs-patterns/lib/fargate/network-multiple-target-groups-fargate-service.ts:109]:
```suggestion
```

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Looks like there is still some logic to be updated with the new fields or I wouldn't have bothered with the neurotic suggestions to fix spacing. But, like, it's all that. This looks good so far. Gotta say, you're really killing it with all these ecs-patterns updates!

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
Huh, I wonder why the appsync error is showing up here. That's been resolved for a while. Maybe it's not printing the most recent build. Anyway, I suspect you've submitted enough PRs recently to have seen this one but, if not, it's a bug we had and then reverted. I would expect it won't show up after this merge.

**Comment by TheRealAmazonKendra** [comment]
Those failures are more like it!

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Same as the other. Just updating the status so I can the alert of the change.

**Comment by michaeldrey** [comment]
Added a few tests for the EC2 side and a readme update. I think this is ready to :shipit:

**Review by TheRealAmazonKendra — APPROVED** [review]
This looks great!

**Pull request merged**